packages/graph-retriever/src/graph_retriever/utils/top_k.py

```diff
@@ -0,0 +1,74 @@
+import heapq
+from typing import Iterable, cast
+
+from graph_retriever.content import Content
+from graph_retriever.utils.math import cosine_similarity_top_k
+
+
+def top_k(batches: Iterable[list[Content]],
+          *,
+          embedding: list[float],
+          k: int,
+          are_batches_sorted: bool = False) -> list[Content]:
+    """
+    Select the top-k contents from the given batches.
+
+    If all contents have scores, they will be used for the comparison rather
+    than being computed.
+
+    Parameters
+    ----------
+    batches : Iterable[list[Content]]
+        The batches of content to select the top-K from.
+    embedding: list[float]
+        The embedding we're looking for.
+    k : int
+        The number of items to select.
+    are_batches_sorted : bool, default False
+        Whether the content of each batch are sorted. If true, and all content
+        has scores, a more efficient top-K selection will be used which doesn't
+        need to consider all of each batch.
+    """
+
+    # TODO: Consider passing threshold here to limit results.
+
+    if all(c.score is not None for batch in batches for c in batch):
+        sorted_items: Iterable[Content]
+        if are_batches_sorted:
+            sorted_items = heapq.merge(*batches, key=_score, reverse=True)
+        else:
+            sorted_items = sorted([c for batch in batches for c in batch],
```

Comment:
This case relies on batches having scores. When that is true, we don't need to do anything with `numpy` for computing similarity. In theory, this case could be triggered by either having the store compute scores in all cases (including when fetching by ID), and in cases where it doesn't we could score just the things that are missing.

Again, not sure on this one.